CIP-XXXX/README.md

```diff
@@ -0,0 +1,205 @@
+     !id   = λ {a} (x :: a) -> x
+in λ (b :: Bool) -> ifThenElse {∀ a. a -> a} b boom id
+```
+Indeed, this example works _without_ adding delays to type abstractions,
+as the untyped form
+```haskell
+f = λ b -> ifThenElse b ERROR (λ x -> x)
+```
+is precisely the example we seek to make work.
+`f True ()` will fail as expected,
+while `f False ()` will return `()` as expected.
+
+### Soundness of optimisations
+
+We note that you can soundly optimise `(\x -> M) N` into `M` if `x` is not mentioned in `M`,
+or similar occurrences where the result of `N` isn't needed.
+Such an optimisation will change the behaviour of the script necessarily,
+as any traces in `N` will no longer happen, yet, this is not important in practice
+as traces are merely a debugging tool. The budget used will decrease,
+but people do not rely on budget exhaustion to signal invalidity.
+
+Transformations like `M || N` into `force (ifThenElse M (delay True) (delay N))` follow a similar principle:
+traces that happened before might no longer happen.
+The used budget will increase if M is `False`, yet this is likely
+still an acceptable transformation for most people.
+
+### Backward compatibility
+
+This CIP has the issue of old scripts not being portable to the new version trivially.
+You need to make sure your script isn't making use of any impure assumptions.
+
+## Path to Active
+
+### Acceptance Criteria
+
+It would be needed to implement the necessary changes in `cardano-ledger` and `plutus`.
+
+### Implementation Plan
+
+#### Implementation of built-ins
+
+This section does not have to be followed and is merely initial speculation
+on how to implement this feature.
+
+The implementation of built-in functions in the Plutus library must
+change to support built-ins that don't inspect all arguments, notably `ifThenElse`.
+The naive way would be to translate errors to `undefined` and catch that impurely, yet this is hardly satisfying.
+
+Built-ins defined by `makeBuiltinMeaning` would, for a Plutus function of type `a0 -> a1 -> ... -> an`,
+be defined by a Haskell-function `forall m. m a0 -> m a1 -> ... m an` instead.
+If you don't bind an argument, it is not used, hence evaluation will still succeed
+regardless of whether it is an error.
+If it is bound, but is an error, then it will return an error as usual.
+
+To reduce the changes needed, you could instead of changing `makeBuiltinMeaning` provide
+an alternative `makeLazyBuiltinMeaning` with the above functionality, changing the original
+function to instead bind all arguments immediately, as most built-ins would do anyway.
+
+With this, the implementation of `ifThenElse` would be
+```haskell
+toBuiltinMeaning DefaultFunVX IfThenElse =
+  makeLazyBuiltinMeaning
+    (\b x y -> b >>= \case True -> x ; False -> y)
+    (runCostingFunThreeArguments . paramIfThenElse)
+```
+
+## Copyright
+
+[CC-BY-4.0]: https://creativecommons.org/licenses/by/4.0/legalcode
+[Apache-2.0]: http://www.apache.org/licenses/LICENSE-2.0
```

Comment:
```suggestion
## Copyright

This CIP is licensed under [CC-BY-4.0].

[CC-BY-4.0]: https://creativecommons.org/licenses/by/4.0/legalcode
[Apache-2.0]: http://www.apache.org/licenses/LICENSE-2.0
```
we do need to choose _one_ of these to display in the markup: I'm going to not suggest Apache since the code herein isn't intended for reuse.